src/main/scala/is/hail/methods/LDPrune.scala

```diff
@@ -0,0 +1,391 @@
+          val gpi = pruneIntermediates(depIndex)
+          pruneIntermediates(depIndex) = gpi.copy(persist = true)
+          (gpi.rdd, (i, gpi.index))
+        }
+      }.unzip
+      (rdds.toArray, inputs.toArray)
+    }
+
+    for (i <- partitionIndices) {
+      val (rdds, inputs) = generalRDDInputs(i)
+      pruneIntermediates(i) = GlobalPruneIntermediate(rdd = new GeneralRDD(sc, rdds, Array((inputs, pruneF))), index = 0, persist = false) // creating single partition RDDs with partition index = 0
+    }
+
+    pruneIntermediates.foreach { gpi =>
+      if (gpi.persist) gpi.rdd.persist(StorageLevel.MEMORY_AND_DISK)
+    }
+
+    val prunedRDD = new GeneralRDD[(Variant, BitPackedVector)](sc, pruneIntermediates.map(_.rdd),
+      pruneIntermediates.zipWithIndex.map { case (gpi, i) =>
+        (Array((i, gpi.index)), pruneF)
+      })
+
+    val annotRDD = prunedRDD.mapValues(_ => Annotation(true)).toOrderedRDD.persist(StorageLevel.MEMORY_AND_DISK)
+    val nVariantsKept = annotRDD.count()
+
+    pruneIntermediates.foreach { gpi => gpi.rdd.unpersist() }
+    inputRDD.unpersist()
+
+    (annotRDD, nVariantsKept)
+  }
+
+  def estimateMemoryRequirements(nVariants: Long, nSamples: Int, nCores: Int, memoryPerCore: Long) = {
+    val nBytesPerVariant = math.ceil(8 * nSamples.toDouble / genotypesPerPack).toLong + variantByteOverhead
+    val memoryAvailPerCore = memoryPerCore * fractionMemoryToUse
+
+    val maxQueueSize = math.max(1, math.ceil(memoryAvailPerCore / nBytesPerVariant).toInt)
+
+    val nPartitionsMinimum = math.max(1, math.ceil(nVariants.toDouble / maxQueueSize).toInt)
+    val nPartitionsOptimal = nCores * nPartitionsPerCore
+
+    val nPartitions =
+      if (nPartitionsOptimal < nPartitionsMinimum)
+        nPartitionsMinimum
+      else
+        nPartitionsOptimal
+
+    assert(maxQueueSize > 0 && nPartitions > 0)
+
+    (maxQueueSize, nPartitions)
+  }
+
+  def apply(vds: VariantDataset, r2Threshold: Double, window: Int, nCores: Int, memoryPerCore: Long = 268435456): VariantDataset = {
+    if (nCores <= 0)
+      fatal(s"Number of cores must be greater than 0.")
+
+    if (r2Threshold < 0 || r2Threshold > 1)
+      fatal(s"R^2 threshold must be in the range [0,1]. Found `$r2Threshold'.")
+
+    if (window < 0)
+      fatal(s"Window must be greater than or equal to 0. Found `$window'.")
+
+    val sc = vds.sparkContext
+
+    val nVariantsInitial = vds.countVariants()
+    val nPartitionsInitial = vds.nPartitions
+    val nSamples = vds.nSamples
+
+    val minMemoryPerCore = math.ceil((1d / fractionMemoryToUse) * 8 * nSamples + variantByteOverhead)
+    val (maxQueueSize, _) = estimateMemoryRequirements(nVariantsInitial, nSamples, nCores, memoryPerCore)
+
+    info(s"InputData: nSamples=$nSamples nVariants=$nVariantsInitial nPartitions=$nPartitionsInitial maxQueueSize=$maxQueueSize")
+
+    if (memoryPerCore < minMemoryPerCore)
+      fatal(s"Memory per core must be greater than ${ minMemoryPerCore / (1024 * 1024) }MB")
```

Comment:
Is there any suggestion you can give to the user on how to proceed if this triggers?